compiler/src/dotty/tools/debug/InsertExpression.scala

```diff
@@ -0,0 +1,238 @@
+        |    unwrapException(constructor.newInstance(args*))
+        |  }
+        |
+        |  def getField(obj: Any, className: String, fieldName: String): Any = {
+        |    val clazz = classLoader.loadClass(className).nn
+        |    val field = clazz.getDeclaredField(fieldName).nn
+        |    field.setAccessible(true)
+        |    field.get(obj)
+        |  }
+        |
+        |  def setField(obj: Any, className: String, fieldName: String, value: Any): Any = {
+        |    val clazz = classLoader.loadClass(className).nn
+        |    val field = clazz.getDeclaredField(fieldName).nn
+        |    field.setAccessible(true)
+        |    field.set(obj, value)
+        |  }
+        |
+        |  def getOuter(obj: Any, outerTypeName: String): Any = {
+        |    val clazz = obj.getClass
+        |    val field = getSuperclassIterator(clazz)
+        |      .flatMap(_.getDeclaredFields.nn.toSeq)
+        |      .map(_.nn)
+        |      .find { field => field.getName == "$$outer" && field.getType.nn.getName == outerTypeName }
+        |      .getOrElse(throw new NoSuchFieldException("$$outer"))
+        |    field.setAccessible(true)
+        |    field.get(obj)
+        |  }
+        |
+        |  def getStaticObject(className: String): Any = {
+        |    val clazz = classLoader.loadClass(className).nn
+        |    val field = clazz.getDeclaredField("MODULE$$").nn
+        |    field.setAccessible(true)
+        |    field.get(null)
+        |  }
+        |
+        |  def getSuperclassIterator(clazz: Class[?] | Null): Iterator[Class[?]] =
+        |    Iterator.iterate(clazz)(_.nn.getSuperclass).takeWhile(_ != null).map(_.nn)
+        |
+        |  // A fake method that is used as a placeholder in the extract-expression phase.
+        |  // The resolve-reflect-eval phase resolves it to a call of one of the other methods in this class.
+        |  def reflectEval(qualifier: Object, term: String, args: Array[Object]): Any = ???
+        |
+        |  private def unwrapException(f: => Any): Any =
+        |    try f catch {
+        |      case e: InvocationTargetException => throw e.getCause.nn
+        |    }
+        |
+        |  extension [T] (x: T | Null) {
+        |    private def nn: T = x.asInstanceOf[T]
+        |  }
+        |}
+        |""".stripMargin
+
+  override def run(using Context): Unit =
+    val inserter = Inserter(parseExpression, parseEvaluationClass)
+    ctx.compilationUnit.untpdTree = inserter.transform(ctx.compilationUnit.untpdTree)
+
+  class Inserter(expression: Tree, expressionClass: Seq[Tree]) extends UntypedTreeMap:
+    override def transform(tree: Tree)(using Context): Tree =
+      tree match
+        case tree: PackageDef =>
+          val transformed = super.transform(tree).asInstanceOf[PackageDef]
+          if expressionInserted then
+            // set to `false` to prevent inserting `Expression` class in other `PackageDef`s
+            expressionInserted = false
```

Comment:
The meaning of this boolean seems backwards to me. Shouldn't it be the other way around:
```suggestion
          if !expressionInserted then
            // set to `true` to prevent inserting `Expression` class in other `PackageDef`s
            expressionInserted = true
```

Reply:
Here we check that we inserted the expression tree in the package def and, if yes, then we insert the expression class.